py/parse_and_notify.py

```diff
@@ -7,51 +7,101 @@
     error_message: str | None,
+    destination: str,
 ) -> str:
     """
-    Creates a markdown table given a sequence of Sequences of cells.
+    Formats an ASCII table for slack message since slack's
+    markdown support is lacking
     """
-
-    def make_row(row: Sequence[str | int | None]) -> str:
-        content = "|".join((str(col) for col in row))
+    if not headers and not content:
+        return ""
+
+    def make_row(
+        row: Sequence[str | int | None], max_width: Sequence[int]
+    ) -> str:
+        content = "|".join(
+            (
+                " " + str(col).ljust(max_width[i]) + " "
+                for i, col in enumerate(row)
+            )
+        )
         return f"|{content}|\n"
 
     assert all(
         len(row) == len(headers) for row in content
     ), "Invalid table format."
+    # get max width of columns
+    num_cols = len(headers)
+    max_width = [len(h) for h in headers]
+    for i in range(num_cols):
+        for row in content:
+            max_width[i] = max(max_width[i], len(str(row[i])))
+
+    # 3 * (len(headers) - 1) because we need to add 3 dashes for each internal
+    # column separator (i.e ' | ')
+    line = ["-" * (width) for width in max_width]
+    rows = [make_row(r, max_width) for r in content]
+    # only create table body if changes actually occurred
+    # (if no changes then `content` = [["Topic Name", {name}]], hence len > 1)
+    table = (
+        (
+            f"{make_row(headers, max_width)}"
+            + f"{make_row(line, max_width)}"
+            + f"{''.join(rows)}"
+        )
+        if len(content) > 1
+        else ""
+    )
+    if destination == Destinations.SLACK and table:
+        table = f"```{table}```"
 
-    line = "-" * len(headers)
-    rows = [make_row(r) for r in content]
-    table = f"{make_row(headers)}{make_row(line)}{''.join(rows)}"
-
-    if error:
+    if error_message is not None:
         error_header = (
-            "# ERROR - the following error occurred while processing this topic:\n"  # noqa
-            f"{error_message}\n\n"
+            "ERROR - the following error occurred while processing this topic:"
         )
-        # if changes were still made before an error, report them
-        if len(content) > 1:
-            table = (
-                error_header
-                + "# The following changes were still made:\n\n"
-                + table
-            )
-        else:
-            table = error_header + "# No changes were made."
+        error_footer = (
+            "The following changes were still made:"
+            if len(content) > 1
+            else "No changes were made."
+        )
+        if destination == Destinations.DATADOG:
+            error_header = f"# {error_header}\n"
+            error_footer = f"# {error_footer}\n"
+        elif destination == Destinations.SLACK:
+            error_header = f":warning: *{error_header}*\n"
+            error_footer = f":warning: *{error_footer}*\n"
```

Comment:
Why do these have to be different ?

Reply:
Slack doesn't support markdown h1's, while datadog markdown doesn't seem to support emojis

I feel like the error text needs to stand out, this was the best way I could come up with